### ISSUE TITLE
Loss function needs more flexibility

### ISSUE THREAD

**Issue by vionwinnie** [lucidrains/mlm-pytorch#2]
In the BERT paper, only the masked tokens prediction are calculted in the crossentropy loss. In the MLM loss calculation, it takes in all the predictions even with the unmasked today. So maybe add attributes to filter out the non-masked label pairs?

**Comment by lucidrains** [comment]
@vionwinnie Hi Winnie! Could you point me to where it says this in the paper? I was under the impression that unmasked tokens need to predict proper logits too, thus why they also swap out some of the tokens with random tokens (~5-10%)

**Comment by lucidrains** [comment]
https://github.com/lucidrains/mlm-pytorch/pull/3 prepared this just in case you are right!

**Comment by vionwinnie** [comment]
Although this allows us to obtain a bidirectional pre-trained model, a downside is that we
are creating a mismatch between pre-training and fine-tuning, since the [MASK] token does not appear during fine-tuning. To mitigate this, we do not always replace “masked” words with the actual [MASK] token. The training data generator chooses 15% of the token positions at random for
prediction. If the i-th token is chosen, we replace the i-th token with (1) the [MASK] token 80% of
the time (2) a random token 10% of the time (3) the unchanged i-th token 10% of the time. Then,
Ti will be used to predict the original token with cross entropy loss. We compare variations of this
procedure in Appendix C.2.

----------------------------------------------------------
Maybe I understand it incorrently too. Still learning about BERT pretraining process :) Thanks for the quick reply!

**Comment by lucidrains** [comment]
im still learning too lol

**Comment by lucidrains** [comment]
@vionwinnie did you figure it out Winnie?

**Comment by lucidrains** [comment]
@vionwinnie Hey Winnie, I believe you were right https://github.com/lucidrains/mlm-pytorch/pull/3 I've merged in the fix. Thank you for reporting this!

**Comment by vionwinnie** [comment]
Thanks. I was gonna reply this morning that I think I was right haha. Looking forward to using your implementation for my proj :+1:
